### ISSUE TITLE
pip install does not work on Python 3

### ISSUE THREAD

**Issue by markkm** [tensorflow/transform#1]
I can confirm that `pip install tensorflow-transform` goes through fine on Python 2.7.13. The following error message shows up for Python 3.5.3 and Python 3.6.0, however:
```
Could not find a version that satisfies the requirement tensorflow-transform (from versions: ) No matching distribution found for tensorflow-transform
```
I see that the package on [PyPI](https://pypi.python.org/pypi/tensorflow-transform) is for Python 2 only. Is this intentional? If so, will there be an upcoming Python 3 release in the near future?

**Comment by KesterTong** [comment]
One of our dependencies, `google-cloud-dataflow`, does not currently support Python 3. There is a [issue](https://github.com/GoogleCloudPlatform/DataflowPythonSDK/issues/6) in their repo tracking this.  As soon as that issue is resolved, we should be able to add Python 3 support for tf.Transform very quickly.

**Comment by carlthome** [comment]
ETA?

**Comment by KesterTong** [comment]
Don't have an ETA at this time, but the Beam developers are actively investigating this issue. In the meantime we will ensure our code is Python 3 ready so that the only remaining issues are the dependencies.

**Comment by kemusa** [comment]
Any update on this for python 3?

**Comment by KesterTong** [comment]
We are still blocked on Beam Python, see https://issues.apache.org/jira/browse/BEAM-1373 for more details

**Comment by jcomfort4** [comment]
Any ETA? It would be helpful if the error message stated this explicitly.  The error message that I got was: 

> Collecting tensorflow-transform
  Could not find a version that satisfies the requirement tensorflow-transform (from versions: )
No matching distribution found for tensorflow-transform

**Comment by KesterTong** [comment]
At this point we can't give an ETA for this as it's blocked by  https://issues.apache.org/jira/browse/BEAM-1373. I know that issue is being discussed but don't have a timeline for it.

Regarding the error message, I will look into this.

**Comment by KesterTong** [comment]
An update: thanks to @cclauss we have #24, #25, #26, #27, #28 which fixed some Python 3 issues in the tf.Transform codebase, though the Beam dependency issue remains.  @cclauss does that fix all the Python 3 issues in tf.Transform code that your'e aware of?

**Comment by cclauss** [comment]
@KesterTong My technique ($ __flake8 . --count --select=E901,E999,F821,F822,F823 --show-source --statistics__) allows me to find and propose fixes to Python 3 porting issues.  We have fixed all the ones that I have found in the tf.Transform codebase but there may be other issues that are not found using my technique.

**Comment by katsiapis** [comment]
As of Tensorflow Transform Release [0.3.1](https://github.com/tensorflow/transform/blob/v0.3.1/RELEASE.md) we offer a Source distribution as well (as opposed to just a WHL).

Going forward (ie for anything >= 0.4), we plan to offer only a Source distribution (and no longer offer a WHL) since Tensorflow Transform is a pure python package.

However as @KesterTong pointed out we still depend on Python 3 support from the Beam SDK (captured by https://issues.apache.org/jira/browse/BEAM-1373).

Beyond the Beam SDK, for folks that want to run Tensorflow Transform on Google Cloud Dataflow, I believe resolution of https://issues.apache.org/jira/browse/BEAM-1251 would also be needed.

/cc @aaltay (in case he can confirm and in case he has any noteworthy updates on the above Jiras).

**Comment by aaltay** [comment]
The previous comment is accurate. @holdenk is making steady progress on the former issue.

**Comment by uriando** [comment]
any update on this ?

**Comment by cclauss** [comment]
* 1 Python 3 Syntax Error in tensorflow/transform
* 14 Python 3 Syntax Errors in apache/beam
Additional work is probably required beyond fixing these Syntax Errors.

flake8 testing of https://github.com/tensorflow/transform on Python 3.6.4

$ __flake8 . --count --select=E901,E999,F821,F822,F823 --show-source --statistics__
```
./tensorflow_transform/beam/impl_test.py:1916:37: E999 SyntaxError: invalid syntax
                         cmp=lambda (xi, xv), (yi, yv): cmp(xv, yv))
                                    ^
```
---
flake8 testing of https://github.com/apache/beam on Python 3.6.4

$ __flake8 . --count --select=E901,E999,F821,F822,F823 --show-source --statistics__
```
    [ ... ]

14    E999 SyntaxError: invalid syntax
115   F821 undefined name 'unicode'
129

**Comment by holdenk** [comment]
@luke-zhu did some excellent work on Python 3 type inference for BEAM (see https://github.com/apache/beam/pull/4183 ) We still have more steps (see https://issues.apache.org/jira/browse/BEAM-2784 ), I have an open PR for a somewhat hacky solution to the proto issue ( https://github.com/apache/beam/pull/4077 ) & more folks are always welcome to join in :)

**Comment by tedmiston** [comment]
Hey @holdenk, is https://issues.apache.org/jira/browse/BEAM-1251 still the best issue to follow for Python 3 support?

Our interest is using Dataflow via Apache Airflow.  I'm trying to find a workaround for that in the short term, like interacting with Dataflow inside a dedicated Py 2 virtual env or Docker container.

**Comment by holdenk** [comment]
I'd say https://issues.apache.org/jira/browse/BEAM-1251 is probably still the right place.

**Comment by redpoint13** [comment]
What the progress and remaining blockers for tft on python3?

**Comment by KesterTong** [comment]
I won't be giving updates on the Apache Beam timeline, please refer to the Apache Beam project.  Apart from Beam dependencies, we have starting running our tests against Python 2 and 3, so we are on track to support Python 3 soon after Apache Beam does.

**Comment by cclauss** [comment]
You might add one more line to your test cases...

[flake8](http://flake8.pycqa.org) testing of https://github.com/tensorflow/transform on Python 3.7.0

$ __flake8 . --count --select=E901,E999,F821,F822,F823 --show-source --statistics__
```
./tensorflow_transform/beam/impl_test.py:2497:14: F821 undefined name 'cmp'
      return cmp(a[1], b[1])
             ^
1     F821 undefined name 'cmp'
1
```

The builtin __cmp()__ was removed in Python 3.

**Comment by rafiqhasan** [comment]
Strange, we are in 2018 with Python 3.x and this still doesn't seem to be a priority. The issue has been open since 2017 and we are soon approaching Python 2.7 end of support.

**Comment by KesterTong** [comment]
@rafiqhasan, Python 3 support is a priority for tf.Transform, but we are blocked by Beam (please see Beam for updates on their Python 3 status).  To re-iterate our strategy from my comment above:
>we have starting running our tests against Python 2 and 3, so we are on track to support Python 3 soon after Apache Beam does.

**Comment by rafiqhasan** [comment]
@KesterTong , highly appreciate your quick comment on it. We are building ML pipelines on Tensorflow and currently all transformations are being managed via our custom transformers that we have attached as functions in the model.

I am pretty sure that most of the real world ML TF projects are already on Python 3+ and involve feature engineering transformations, so how is this unavailability of Py 3 support being tackled in those projects ? Is no one else using TFT in production with Py 3 ? Are there any alternatives ?

**Comment by KesterTong** [comment]
@rafiqhasan One example of users who are able to user TFT in production, is existing users of Apache Beam or Google Cloud Dataflow, since they would already by using Python 2.  We will be adding Python 3 support once https://issues.apache.org/jira/browse/BEAM-1251 is fixed.

**Comment by rafiqhasan** [comment]
> Is this still an issue ?

Yes it is and should stay until https://issues.apache.org/jira/browse/BEAM-1251 is fixed.

**Comment by Selimonder** [comment]
is this still an issue?

**Comment by cclauss** [comment]
https://pypi.org/project/tensorflow-transform/#data still says that tf-transform only support legacy Python with just one year left until it’s end-of-life...  http://pythonclock.org

**Comment by impredicative** [comment]
Google will probably have to get involved a lot more closely in the Beam project to address this. Google-specific issues with [DataflowPythonSDK](https://github.com/GoogleCloudPlatform/DataflowPythonSDK/issues/6) and [datalab](https://github.com/googledatalab/datalab/issues/2046) also look to be affected, motivating a need for closer involvement.

**Comment by holdenk** [comment]
There’s been some good progress, see https://issues.apache.org/jira/browse/BEAM-1251 , but it does seem we are still not there yet.

**Comment by xxllp** [comment]
too long time ~

**Comment by katsiapis** [comment]
TFT 0.13 has support for Python 3.5 [1].

Other version of Python (like 3.6 and 3.7) will follow in future releases (after Beam officially supports them).

[1] https://github.com/tensorflow/transform/releases/tag/v0.13.0

**Comment by katsiapis** [comment]
@zoyahav Should we close this now?

**Comment by cclauss** [comment]
#101 seems to be related.

**Comment by zoyahav** [comment]
This library currently supports python 2.7, 3.5-3.7 (since tf.transform 0.14).
If there are any issues with tensorflow_transform>=0.14 and python 3 please open a new issue, I think it's safe to close this one.
